File: backend/cambridge.py
import requests
from bs4 import BeautifulSoup

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"
}

def get_cambridge_audio(word: str):
    """
    Scrapes Cambridge Dictionary for the first available audio URL for a given word.
    Prioritizes US English if available, otherwise UK.
    """
    try:
        url = f"https://dictionary.cambridge.org/dictionary/english/{word}"
        r = requests.get(url, headers=HEADERS, timeout=5)
        if r.status_code != 200:
            return None
        
        doc = BeautifulSoup(r.text, "html.parser")
        
        # Strategy: Look for audio sources
        # We prefer "us" or "uk"
        
        audio_sources = []
        
        for source in doc.select("audio source[src]"):
            src = source.get("src")
            if src:
                if src.startswith("/"):
                    src = "https://dictionary.cambridge.org" + src
                audio_sources.append(src)
                
        # Also check buttons
        for btn in doc.select(".audio_play_button, .audio"):
            mp3 = btn.get("data-src-mp3") or btn.get("data-src")
            if mp3:
                 if mp3.startswith("/"):
                    mp3 = "https://dictionary.cambridge.org" + mp3
                 audio_sources.append(mp3)

        if not audio_sources:
            return None

        # Prefer US English (.mp3 and contains 'us_')
        for src in audio_sources:
            if "us_" in src and src.endswith(".mp3"):
                return src
        
        # Fallback to any MP3
        for src in audio_sources:
            if src.endswith(".mp3"):
                return src
                
        return audio_sources[0] if audio_sources else None
        
    except Exception as e:
        print(f"Cambridge scrape error for {word}: {e}")
        return None

File: backend/cambridge.js
import axios from 'axios';
import * as cheerio from 'cheerio';

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
};

export async function getCambridgeAudio(word) {
  try {
    const url = `https://dictionary.cambridge.org/dictionary/english/${word}`;
    const { data } = await axios.get(url, { headers: HEADERS, timeout: 5000 });
    
    const $ = cheerio.load(data);
    const audioSources = [];

    // Strategy 1: <audio> <source src="...">
    $('audio source[src]').each((_, el) => {
      let src = $(el).attr('src');
      if (src) {
        if (src.startsWith('/')) src = `https://dictionary.cambridge.org${src}`;
        audioSources.push(src);
      }
    });

    // Strategy 2: Buttons with data-src
    $('.audio_play_button, .audio').each((_, el) => {
      let src = $(el).attr('data-src-mp3') || $(el).attr('data-src');
      if (src) {
        if (src.startsWith('/')) src = `https://dictionary.cambridge.org${src}`;
        audioSources.push(src);
      }
    });

    if (audioSources.length === 0) return null;

    // Filter Logic
    // 1. Prefer US English (us_) and .mp3
    const usMp3 = audioSources.find(s => s.includes('us_') && s.endsWith('.mp3'));
    if (usMp3) return usMp3;

    // 2. Fallback to any MP3
    const anyMp3 = audioSources.find(s => s.endsWith('.mp3'));
    if (anyMp3) return anyMp3;

    return audioSources[0];
  } catch (error) {
    console.error(`Cambridge scrape error for ${word}:`, error.message);
    return null;
  }
}
